src/main/resources/dbreplica/migration/V1.0.4__reportable_task_plsql.sql

```diff
@@ -0,0 +1,202 @@
+--
+-- Copies a task record into the history table.
+--
+CREATE OR REPLACE FUNCTION cft_task_db.add_reportable_task(l_task cft_task_db.tasks)
+ RETURNS bigint
+ LANGUAGE plpgsql
+AS $function$
+declare
+  l_update_id bigint;
+  l_new_task boolean;
+  l_result_action varchar default 'ignore';
+  l_update_lock integer;
+
+begin
+
+  -- checking relevant fields are populated relating to a new task being assigned
+select not exists(select 1 from cft_task_db.reportable_task where reportable_task.task_id = l_task.task_id)
+intersect
+select not exists(select 1
+                  from cft_task_db.task_history
+                  where task_history.task_id = l_task.task_id
+                  having count(*) > 1)
+into l_new_task;
+
+if (l_task.last_reconfiguration_time is null) and
+      ((l_task.last_updated_action = 'Configure' and l_task.state in ('UNASSIGNED', 'ASSIGNED'))
+        or (l_task.last_updated_action = 'AutoAssign' and l_task.state = 'ASSIGNED')) then
+    l_result_action = 'insert';
+
+    if not l_new_task then
+          l_result_action = 'update';
+
+    end if;
+end if;
+
+if (l_result_action = 'ignore') and not l_new_task then
+    l_result_action = 'update';
+end if;
+
+if l_result_action = 'insert' then
+    insert into cft_task_db.reportable_task
+    (task_id, task_name, task_type, due_date_time,
+     state, task_system, security_classification,
+     title, major_priority, minor_priority, assignee,
+     auto_assigned, execution_type_code, work_type,
+     role_category, has_warnings, assignment_expiry,
+     case_id, case_type_id, case_category, case_name,
+     jurisdiction, region, location, business_context,
+     termination_reason, created, updated_by, updated,
+     update_action, created_date, due_date, last_updated_date)
+    values
+      (l_task.task_id, l_task.task_name, l_task.task_type, l_task.due_date_time,
+      l_task.state, l_task.task_system, l_task.security_classification,
+      l_task.title, l_task.major_priority, l_task.minor_priority, l_task.assignee,
+      l_task.auto_assigned, l_task.execution_type_code, l_task.work_type,
+      l_task.role_category, l_task.has_warnings, l_task.assignment_expiry,
+      l_task.case_id, l_task.case_type_id, l_task.case_category, l_task.case_name,
+      l_task.jurisdiction, l_task.region, l_task.location, l_task.business_context,
+      l_task.termination_reason, l_task.created, l_task.last_updated_user, l_task.last_updated_timestamp,
+      l_task.last_updated_action, l_task.created::DATE, l_task.due_date_time::date,
+      l_task.last_updated_timestamp::date)
+    returning update_id into l_update_id;
+end if;
+
+if l_result_action = 'update' then
+    select 1 into l_update_lock from cft_task_db.reportable_task
+    where reportable_task.task_id = l_task.task_id FOR UPDATE;
```

Comment:
Probably need to talk about this a bit.  What happens in the case that the task_history has multiple rows but there is no record in the reportable_task table?  Looks like here we are only making an update/insert based on the current update?  Does that work in the above scenario?